client/driver/executor/checks.go

```diff
@@ -0,0 +1,174 @@
+package executor
+
+import (
+	"fmt"
+	"log"
+	"os/exec"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/armon/circbuf"
+	docker "github.com/fsouza/go-dockerclient"
+
+	cstructs "github.com/hashicorp/nomad/client/driver/structs"
+)
+
+var (
+	// We store the client globally to cache the connection to the docker daemon.
+	createClient sync.Once
+	client       *docker.Client
+)
+
+// DockerScriptCheck runs nagios compatible scripts in a docker container and
+// provides the check result
+type DockerScriptCheck struct {
+	id          string
+	interval    time.Duration
+	containerID string
+	logger      *log.Logger
+	cmd         string
+	args        []string
+
+	dockerEndpoint string
+	tlsCert        string
+	tlsCa          string
+	tlsKey         string
+}
+
+// dockerClient creates the client to interact with the docker daemon
+func (d *DockerScriptCheck) dockerClient() (*docker.Client, error) {
+	if client != nil {
+		return client, nil
+	}
+
+	var err error
+	createClient.Do(func() {
+		if d.dockerEndpoint != "" {
+			if d.tlsCert+d.tlsKey+d.tlsCa != "" {
+				d.logger.Printf("[DEBUG] executor.checks: using TLS client connection to %s", d.dockerEndpoint)
+				client, err = docker.NewTLSClient(d.dockerEndpoint, d.tlsCert, d.tlsKey, d.tlsCa)
+			} else {
+				d.logger.Printf("[DEBUG] executor.checks: using standard client connection to %s", d.dockerEndpoint)
+				client, err = docker.NewClient(d.dockerEndpoint)
+			}
+			return
+		}
+
+		d.logger.Println("[DEBUG] executor.checks: using client connection initialized from environment")
+		client, err = docker.NewClientFromEnv()
+	})
+	return client, err
+}
+
+// Run runs a script check inside a docker container
+func (d *DockerScriptCheck) Run() *cstructs.CheckResult {
+	var (
+		exec    *docker.Exec
+		err     error
+		execRes *docker.ExecInspect
+		time    = time.Now()
+	)
+
+	if client, err = d.dockerClient(); err != nil {
+		return &cstructs.CheckResult{Err: err}
+	}
+	client = client
+	execOpts := docker.CreateExecOptions{
+		AttachStdin:  false,
+		AttachStdout: true,
+		AttachStderr: true,
+		Tty:          false,
+		Cmd:          append([]string{d.cmd}, d.args...),
+		Container:    d.containerID,
+	}
+	if exec, err = client.CreateExec(execOpts); err != nil {
+		return &cstructs.CheckResult{Err: err}
+	}
+
+	output, _ := circbuf.NewBuffer(int64(cstructs.CheckBufSize))
+	startOpts := docker.StartExecOptions{
+		Detach:       false,
+		Tty:          false,
+		OutputStream: output,
+		ErrorStream:  output,
+	}
+
+	if err = client.StartExec(exec.ID, startOpts); err != nil {
+		return &cstructs.CheckResult{Err: err}
+	}
+	if execRes, err = client.InspectExec(exec.ID); err != nil {
```

Comment:
@dadgar StartExec starts the exec process asynchronously and inspectexec should return too.

But the daemon can be un-responsive and can block indefinitely like any other calls to docker. And the docker client doesn't have the notion of timing out if the daemon takes a really long time.